File: droid/app/src/main/java/com/storyteller/droid/engine/ChunkStream.kt
package com.storyteller.droid.engine

import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.receiveAsFlow

/**
 * P8.6 — Frozen stream event types, identical across Python, Kotlin, Swift.
 *
 * One request owns one model lease and cancellation token.  Cancellation
 * emits [CANCELLED] — never [FAILED].
 */
sealed class ChunkStreamEvent {
    abstract val requestId: String
    abstract val eventType: String

    data class Started(override val requestId: String) : ChunkStreamEvent() {
        override val eventType: String get() = "started"
    }

    data class Text(
        override val requestId: String,
        val sequence: Int,
        val text: String,
    ) : ChunkStreamEvent() {
        override val eventType: String get() = "text"
        init { require(text.isNotEmpty()) { "text chunk must be non-empty" } }
    }

    data class Completed(
        override val requestId: String,
        val sequence: Int,
        val usage: Map<String, Int>,
    ) : ChunkStreamEvent() {
        override val eventType: String get() = "completed"
    }

    data class Failed(
        override val requestId: String,
        val sequence: Int,
        val stableCode: String,
    ) : ChunkStreamEvent() {
        override val eventType: String get() = "failed"
    }

    data class Cancelled(
        override val requestId: String,
        val sequence: Int,
        val stableCode: String = "STREAM_CANCELLED",
    ) : ChunkStreamEvent() {
        override val eventType: String get() = "cancelled"
    }

    /** Every terminal event carries a stable diagnostic code. */
    val terminalCode: String?
        get() = when (this) {
            is Failed -> stableCode
            is Cancelled -> stableCode
            is Completed -> null
            else -> null
        }
}

/**
 * P8.6 — Bounded channel between native callbacks and UI consumer.
 *
 * Fixed capacity ([DEFAULT_CAPACITY] = 64). [send] suspends when the producer
 * outruns the consumer. Events are never dropped, rewritten, or merged.
 *
 * One [BoundedChunkChannel] per request.  Close it to signal the consumer.
 */
class BoundedChunkChannel(capacity: Int = DEFAULT_CAPACITY) {
    companion object {
        const val DEFAULT_CAPACITY = 64
        const val MIN_CAPACITY = 4
    }

    init { require(capacity >= MIN_CAPACITY) { "capacity must be at least $MIN_CAPACITY" } }

    private val channel = Channel<ChunkStreamEvent>(capacity)

    fun events(): Flow<ChunkStreamEvent> = channel.receiveAsFlow()

    suspend fun send(event: ChunkStreamEvent) {
        channel.send(event)
    }

    fun close() { channel.close() }
}

/** P8.C2 stable diagnostic codes for stream errors. */
object StreamErrorCodes {
    const val MODEL_NOT_LOADED = "STREAM_MODEL_NOT_LOADED"
    const val CANCELLED = "STREAM_CANCELLED"
    const val NATIVE_FAILURE = "STREAM_NATIVE_FAILURE"
    const val TIMEOUT = "STREAM_TIMEOUT"
    const val QUEUE_FULL = "STREAM_QUEUE_FULL"
}

/**
 * P8.6 — Builds a correctly-ordered sequence of [ChunkStreamEvent]s.
 *
 * Usage:
 * ```
 * val builder = StreamBuilder("req_01")
 * channel.send(builder.started())
 * channel.send(builder.text("Hello "))
 * channel.send(builder.completed(mapOf("tokens" to 2)))
 * ```
 */
class StreamBuilder(val requestId: String) {
    private var seq = 0

    fun started() = ChunkStreamEvent.Started(requestId)

    fun text(text: String): ChunkStreamEvent.Text {
        seq++
        return ChunkStreamEvent.Text(requestId, seq, text)
    }

    fun completed(usage: Map<String, Int>): ChunkStreamEvent.Completed {
        seq++
        return ChunkStreamEvent.Completed(requestId, seq, usage)
    }

    fun failed(code: String): ChunkStreamEvent.Failed {
        seq++
        return ChunkStreamEvent.Failed(requestId, seq, code)
    }

    fun cancelled(): ChunkStreamEvent.Cancelled {
        seq++
        return ChunkStreamEvent.Cancelled(requestId, seq)
    }
}

File: ios/StoryTeller/Engine/ChunkStream.swift
import Foundation

// MARK: - P8.6 Chunk Stream Events

/// Frozen stream event types — identical across Python, Kotlin, Swift.
///
/// One request owns one model lease and cancellation token.  Cancellation
/// emits ``cancelled`` — never ``failed``.
enum ChunkStreamEventType: String, Sendable {
    case started, text, completed, failed, cancelled
}

/// One emitted event from the chunk stream.
///
/// All fields are stable across platforms.  `requestId` ties events to their
/// originating request.  `sequence` is strictly increasing within a request
/// and starts at 0 for STARTED.
struct ChunkStreamEvent: Sendable {
    let requestId: String
    let eventType: ChunkStreamEventType
    let sequence: Int
    let text: String
    let usage: [String: Int]?
    let stableCode: String

    init(requestId: String, eventType: ChunkStreamEventType, sequence: Int,
         text: String = "", usage: [String: Int]? = nil, stableCode: String = "") {
        self.requestId = requestId
        self.eventType = eventType
        self.sequence = sequence
        self.text = text
        self.usage = usage
        self.stableCode = stableCode
    }

    var terminalCode: String? {
        switch eventType {
        case .failed, .cancelled: return stableCode.isEmpty ? nil : stableCode
        default: return nil
        }
    }
}

// MARK: - P8.C2 Stable Diagnostic Codes

enum StreamErrorCode {
    static let modelNotLoaded = "STREAM_MODEL_NOT_LOADED"
    static let cancelled = "STREAM_CANCELLED"
    static let nativeFailure = "STREAM_NATIVE_FAILURE"
    static let timeout = "STREAM_TIMEOUT"
    static let queueFull = "STREAM_QUEUE_FULL"
}

// MARK: - Bounded Async Channel

/// P8.6 bounded, lossless bridge from synchronous native callbacks to async UI.
///
/// `send` applies backpressure when 64 events are pending. No token text is
/// dropped, rewritten, or merged. One channel has one consumer and one terminal.
final class BoundedChunkChannel: AsyncSequence, @unchecked Sendable {
    typealias Element = ChunkStreamEvent

    struct AsyncIterator: AsyncIteratorProtocol {
        fileprivate let channel: BoundedChunkChannel
        mutating func next() async -> ChunkStreamEvent? { await channel.nextEvent() }
    }

    static let defaultCapacity = 64
    static let minCapacity = 4

    private let capacity: Int
    private let condition = NSCondition()
    private var buffer: [ChunkStreamEvent] = []
    private var waiter: CheckedContinuation<ChunkStreamEvent?, Never>?
    private var cancellationHandler: (@Sendable () -> Void)?
    private var finished = false

    init(capacity: Int = defaultCapacity) {
        precondition(capacity >= Self.minCapacity, "capacity must be at least \(Self.minCapacity)")
        self.capacity = capacity
    }

    func makeAsyncIterator() -> AsyncIterator { AsyncIterator(channel: self) }

    func onCancellation(_ handler: @escaping @Sendable () -> Void) {
        condition.lock()
        cancellationHandler = handler
        condition.unlock()
    }

    func send(_ event: ChunkStreamEvent) {
        condition.lock()
        while buffer.count >= capacity && waiter == nil && !finished {
            condition.wait()
        }
        guard !finished else { condition.unlock(); return }
        let waiting = waiter
        waiter = nil
        if waiting == nil { buffer.append(event) }
        if event.eventType == .completed || event.eventType == .failed || event.eventType == .cancelled {
            finished = true
        }
        condition.broadcast()
        condition.unlock()
        waiting?.resume(returning: event)
    }

    func finish() {
        condition.lock()
        finished = true
        let waiting = waiter
        waiter = nil
        condition.broadcast()
        condition.unlock()
        waiting?.resume(returning: nil)
    }

    private func nextEvent() async -> ChunkStreamEvent? {
        await withTaskCancellationHandler {
            await withCheckedContinuation { continuation in
                condition.lock()
                if !buffer.isEmpty {
                    let event = buffer.removeFirst()
                    condition.signal()
                    condition.unlock()
                    continuation.resume(returning: event)
                } else if finished {
                    condition.unlock()
                    continuation.resume(returning: nil)
                } else {
                    precondition(waiter == nil, "BoundedChunkChannel supports one consumer")
                    waiter = continuation
                    condition.unlock()
                }
            }
        } onCancel: {
            self.cancelConsumer()
        }
    }

    private func cancelConsumer() {
        condition.lock()
        let handler = cancellationHandler
        cancellationHandler = nil
        condition.unlock()
        handler?()
        finish()
    }
}

// MARK: - Stream Builder

/// P8.6 — Builds a correctly-ordered sequence of `ChunkStreamEvent`s.
struct StreamBuilder {
    let requestId: String
    private var seq = 0

    init(requestId: String) {
        self.requestId = requestId
    }

    func started() -> ChunkStreamEvent {
        ChunkStreamEvent(requestId: requestId, eventType: .started, sequence: 0)
    }

    mutating func text(_ text: String) -> ChunkStreamEvent {
        precondition(!text.isEmpty, "text chunk must be non-empty")
        seq += 1
        return ChunkStreamEvent(requestId: requestId, eventType: .text, sequence: seq, text: text)
    }

    mutating func completed(_ usage: [String: Int]) -> ChunkStreamEvent {
        seq += 1
        return ChunkStreamEvent(requestId: requestId, eventType: .completed, sequence: seq, usage: usage)
    }

    mutating func failed(_ code: String) -> ChunkStreamEvent {
        seq += 1
        return ChunkStreamEvent(requestId: requestId, eventType: .failed, sequence: seq, stableCode: code)
    }

    mutating func cancelled() -> ChunkStreamEvent {
        seq += 1
        return ChunkStreamEvent(requestId: requestId, eventType: .cancelled, sequence: seq, stableCode: StreamErrorCode.cancelled)
    }
}
